File: src/ui/modules/montr-core/models/api-result.ts
import { Guid } from ".";

export interface ApiResult<TData = any> {
	success: boolean;
	message?: string;
	uid?: Guid;
	affectedRows?: number;
	concurrencyStamp?: string;
	errors?: ApiResultError[];
	redirectUrl?: string;
	data?: TData;
}

export interface ApiResultError {
	key: string;
	messages: string[];
}

export interface ProblemDetails {
	detail?: string;
	instance?: string;
	status?: number;
	title?: string;
	type?: string;
}

export interface ValidationProblemDetails extends ProblemDetails {
	errors?: Map<string, string[]>;
}

File: src/Montr.Core/Models/ApiResult.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Montr.Core.Models
{
	public class ApiResult
	{
		public ApiResult()
		{
			Success = true;	
		}

		public bool Success { get; set; }

		/// <summary>
		/// Explanatory message for result.</summary>
		public string Message { get; set; }

		/// <summary>
		/// Uid of new row in case of insert operation.</summary>
		public Guid? Uid { get; set; }

		public long? AffectedRows { get; set; }

		public string RedirectUrl { get; set; }

		public IList<ApiResultError> Errors { get; set; }

		// todo: add tests
		public void AssertSuccess(Func<string> getMessage)
		{
			if (Success == false)
			{
				var message = getMessage();

				// todo: get detailed errors from result
				var errors = string.Join(", ", Errors.SelectMany(x => x.Messages));

				// todo: use ApiException ?
				throw new ApplicationException($"{message} - {errors}");
			}
		}
	}

	public class ApiResult<TData> : ApiResult
	{
		public TData Data { get; set; }
	}

	public class ApiResultError
	{
		public string Key { get; set; }

		public string[] Messages { get; set; }
	}
}
